File: aidl/hwcontrol/java_extension/src/com/harshit/nabuextensions/PeripheralUtils.java
//
// Copyright (C) 2023 Harshit Jain <dev-harsh1998@hotmail.com>
//
// SPDX-License-Identifier: Apache-2.0
//
package com.harshit.nabuextensions;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.UserHandle;
import android.os.SystemProperties;
import android.util.Log;

import com.harshit.nabuextensions.HwStateManager;
import custom.hardware.hwcontrol.HwType;

import static com.harshit.nabuextensions.stylus.StylusSettingsFragment.SHARED_STYLUS;
import static com.harshit.nabuextensions.stylus.StylusSettingsFragment.SHARED_STYLUS_GEN;
import static com.harshit.nabuextensions.stylus.StylusSettingsFragment.SHARED_STYLUS_GEN_PROP;
import static com.harshit.nabuextensions.keyboard.XiaomiKeyboardSettingsFragment.SHARED_KEYBOARD;
import static com.harshit.nabuextensions.tap2wake.Tap2WakeSettingsFragment.SHARED_TAP2WAKE;

public class PeripheralUtils {
    private static final String TAG = "PeripheralUtils";
    private static final boolean DEBUG = false;
    private static HwStateManager mHwStateManager;
    private static SharedPreferences stylus;
    private static SharedPreferences keyboard;
    private static SharedPreferences tap2wake;
    private static SharedPreferences stylusGen;

    public static void BootResetState(Context context) {
        if (DEBUG)
            Log.d(TAG, "Starting service");
        // Initialize stylus and keyboard shared preferences
        stylus = context.getSharedPreferences(SHARED_STYLUS, Context.MODE_PRIVATE);
        keyboard = context.getSharedPreferences(SHARED_KEYBOARD, Context.MODE_PRIVATE);
        tap2wake = context.getSharedPreferences(SHARED_TAP2WAKE, Context.MODE_PRIVATE);
        stylusGen = context.getSharedPreferences(SHARED_STYLUS_GEN, Context.MODE_PRIVATE);

        // Initialize HwStateManager
        HwStateManager.BootResetState();

        // Sync all peripherals
        SyncAll();
    }

    // Enable stylus based on shared preference.
    private static void SyncStylus() {
        if (DEBUG)
            Log.d(TAG, "Enabling stylus");
        mHwStateManager.HwState(HwType.STYLUS, stylus.getInt(SHARED_STYLUS, 0));
        SystemProperties.set(SHARED_STYLUS_GEN_PROP, stylusGen.getString(SHARED_STYLUS_GEN, "1"));
    }

    // Enable keyboard based on shared preference.
    private static void SyncKeyboard() {
        if (DEBUG)
            Log.d(TAG, "Enabling keyboard");
        mHwStateManager.HwState(HwType.KEYBOARD, keyboard.getInt(SHARED_KEYBOARD, 0));
    }

    // Enable tap2wake based on shared preference.
    public static void SyncTap2Wake() {
        if (DEBUG)
            Log.d(TAG, "Enabling tap2wake");
        mHwStateManager.HwState(HwType.TAP2WAKE, tap2wake.getInt(SHARED_TAP2WAKE, 0));
    }

    // Sync all peripherals
    private static void SyncAll() {
        SyncStylus();
        SyncKeyboard();
        SyncTap2Wake();
    }
}

File: aidl/hwcontrol/java_extension/src/com/harshit/nabuextensions/PeripheralUtils.kt
/*
 * Copyright (C) 2023 Harshit Jain <dev-harsh1998@hotmail.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package com.harshit.nabuextensions

import android.content.Context
import android.os.SystemProperties
import android.util.Log
import custom.hardware.hwcontrol.HwType

/**
 * Utility class for managing peripheral devices (stylus, keyboard, tap2wake).
 * Handles synchronization of hardware states on boot and preference changes.
 */
object PeripheralUtils {
    
    private val TAG = PeripheralUtils::class.java.simpleName
    
    /**
     * Initialize all peripherals on boot completion.
     * Restores saved preferences and applies hardware states.
     */
    fun bootResetState(context: Context) {
        Log.i(TAG, "Initializing peripheral devices")
        
        // Initialize hardware control manager
        HwStateManager.initialize()
        
        // Restore all peripheral states from preferences
        syncAll(context)
        
        Log.i(TAG, "Peripheral initialization complete")
    }
    
    /**
     * Synchronize stylus hardware state with saved preferences
     */
    fun syncStylus(context: Context) {
        val enabled = PreferenceManager.getStylusEnabled(context)
        val generation = PreferenceManager.getStylusGeneration(context)
        
        Log.d(TAG, "Syncing stylus: enabled=$enabled, generation=$generation")
        
        HwStateManager.setHwState(HwType.STYLUS, enabled)
        SystemProperties.set(PreferenceManager.KEY_STYLUS_GEN_PROP, generation)
    }
    
    /**
     * Synchronize keyboard hardware state with saved preferences
     */
    fun syncKeyboard(context: Context) {
        val enabled = PreferenceManager.getKeyboardEnabled(context)
        
        Log.d(TAG, "Syncing keyboard: enabled=$enabled")
        
        HwStateManager.setHwState(HwType.KEYBOARD, enabled)
    }
    
    /**
     * Synchronize tap2wake hardware state with saved preferences
     */
    fun syncTap2Wake(context: Context) {
        val enabled = PreferenceManager.getTap2WakeEnabled(context)
        
        Log.d(TAG, "Syncing tap2wake: enabled=$enabled")
        
        HwStateManager.setHwState(HwType.TAP2WAKE, enabled)
    }
    
    /**
     * Synchronize game mode hardware state with saved preferences
     */
    fun syncGameMode(context: Context) {
        val enabled = PreferenceManager.getGameModeEnabled(context)
        
        Log.d(TAG, "Syncing game mode: enabled=$enabled")
        
        HwStateManager.setHwState(HwType.GAMEMODE, enabled)
    }
    
    /**
     * Synchronize all peripheral hardware states
     */
    private fun syncAll(context: Context) {
        syncStylus(context)
        syncKeyboard(context)
        syncTap2Wake(context)
        syncGameMode(context)
    }
}

